sync tries https for sftp

```
root@host ~ # juicefs -v sync  sftp://user:pass@user.rsync.net/juicecloud b2://xxxx:yyy/cloudstorage
2023/12/19 14:57:30.030460 juicefs[672723] <DEBUG>: maxprocs: Leaving GOMAXPROCS=4: CPU quota undefined [maxprocs.go:47]
2023/12/19 14:57:30.031713 juicefs[672723] <DEBUG>: Creating sftp storage at endpoint https://user.rsync.net [object_storage.go:171]
2023/12/19 14:57:30.033716 juicefs[672723] <ERROR>: connect to https failed: dial tcp: lookup https: no such host [sftp.go:539]
2023/12/19 14:57:30.033841 juicefs[672723] <FATAL>: create sftp https://user.rsync.net: dial tcp: lookup https: no such host [main.go:31]
```

**Environment**:
- JuiceFS version (use `juicefs --version`) or Hadoop Java SDK version: 1.1.1
- Cloud provider or hardware configuration running JuiceFS: rsync.net / b2
- OS (e.g `cat /etc/os-release`): Arch Linux
- Kernel (e.g. `uname -a`): 6.1.68
- Object storage (cloud provider and region, or self maintained):
- Metadata engine info (version, cloud provider managed or self maintained): redis
- Network connectivity (JuiceFS to metadata engine, JuiceFS to object storage): local

The workaround leads to another issue:
```
2023/12/20 15:41:04.883244 juicefs[1115587] <DEBUG>: Creating sftp storage at endpoint user.rsync.net/juicecloud [object_storage.go:171]
panic: runtime error: slice bounds out of range [:-1]
```

--- a/cmd/sync.go
+++ b/cmd/sync.go
@@ -279,6 +279,7 @@ func extractToken(uri string) (string, string) {
 }
 
 func createSyncStorage(uri string, conf *sync.Config) (object.ObjectStorage, error) {
+	// nolint:staticcheck
 	uri = strings.TrimLeft(uri, "sftp://")
 	if !strings.Contains(uri, "://") {
 		if isFilePath(uri) {

--- a/cmd/sync.go
+++ b/cmd/sync.go
@@ -279,9 +279,7 @@ func extractToken(uri string) (string, string) {
 }
 
 func createSyncStorage(uri string, conf *sync.Config) (object.ObjectStorage, error) {
-	if strings.HasPrefix(uri, "sftp://") {
-		uri = strings.TrimLeft(uri, "sftp://")
-	}
+	uri = strings.TrimLeft(uri, "sftp://")
 	if !strings.Contains(uri, "://") {
 		if isFilePath(uri) {
 			absPath, err := filepath.Abs(uri)

--- a/cmd/sync.go
+++ b/cmd/sync.go
@@ -295,19 +295,7 @@ func createSyncStorage(uri string, conf *sync.Config) (object.ObjectStorage, err
 			// Windows: file:///C:/a/b/c, Unix: file:///a/b/c
 			uri = "file://" + absPath
 		} else { // sftp
-			var user string
-			if strings.Contains(uri, "@") {
-				parts := strings.Split(uri, "@")
-				user = parts[0]
-				uri = parts[1]
-			}
-			var pass string
-			if strings.Contains(user, ":") {
-				parts := strings.Split(user, ":")
-				user = parts[0]
-				pass = parts[1]
-			}
-			return object.CreateStorage("sftp", uri, user, pass, "")
+			uri = "sftp://" + uri
 		}
 	}
 	uri, token := extractToken(uri)
@@ -338,6 +326,10 @@ func createSyncStorage(uri string, conf *sync.Config) (object.ObjectStorage, err
 		}
 	} else if name == "nfs" {
 		endpoint = u.Host + u.Path
+	} else if name == "sftp" {
+		if strings.Contains(uri, "@") {
+			endpoint = strings.Split(uri, "@")[1]
+		}
 	} else if !conf.NoHTTPS && supportHTTPS(name, u.Host) {
 		endpoint = "https://" + u.Host
 	} else {
@@ -369,7 +361,7 @@ func createSyncStorage(uri string, conf *sync.Config) (object.ObjectStorage, err
 		}
 	}
 	switch name {
-	case "file", "nfs":
+	case "file", "nfs", "sftp":
 	case "minio":
 		if strings.Count(u.Path, "/") > 1 {
 			// skip bucket name

--- a/cmd/sync.go
+++ b/cmd/sync.go
@@ -295,7 +295,19 @@ func createSyncStorage(uri string, conf *sync.Config) (object.ObjectStorage, err
 			// Windows: file:///C:/a/b/c, Unix: file:///a/b/c
 			uri = "file://" + absPath
 		} else { // sftp
-			uri = "sftp://" + uri
+			var user string
+			if strings.Contains(uri, "@") {
+				parts := strings.Split(uri, "@")
+				user = parts[0]
+				uri = parts[1]
+			}
+			var pass string
+			if strings.Contains(user, ":") {
+				parts := strings.Split(user, ":")
+				user = parts[0]
+				pass = parts[1]
+			}
+			return object.CreateStorage("sftp", uri, user, pass, "")
 		}
 	}
 	uri, token := extractToken(uri)
@@ -326,10 +338,6 @@ func createSyncStorage(uri string, conf *sync.Config) (object.ObjectStorage, err
 		}
 	} else if name == "nfs" {
 		endpoint = u.Host + u.Path
-	} else if name == "sftp" {
-		if strings.Contains(uri, "@") {
-			endpoint = strings.Split(uri, "@")[1]
-		}
 	} else if !conf.NoHTTPS && supportHTTPS(name, u.Host) {
 		endpoint = "https://" + u.Host
 	} else {
@@ -361,7 +369,7 @@ func createSyncStorage(uri string, conf *sync.Config) (object.ObjectStorage, err
 		}
 	}
 	switch name {
-	case "file", "nfs", "sftp":
+	case "file", "nfs":
 	case "minio":
 		if strings.Count(u.Path, "/") > 1 {
 			// skip bucket name

--- a/cmd/sync.go
+++ b/cmd/sync.go
@@ -279,6 +279,9 @@ func extractToken(uri string) (string, string) {
 }
 
 func createSyncStorage(uri string, conf *sync.Config) (object.ObjectStorage, error) {
+	if strings.HasPrefix(uri, "sftp://") {
+		uri = strings.TrimLeft(uri, "sftp://")
+	}
 	if !strings.Contains(uri, "://") {
 		if isFilePath(uri) {
 			absPath, err := filepath.Abs(uri)

